app/src/lib/stores/helpers/branch-pruner.ts

```diff
@@ -0,0 +1,153 @@
+import { Repository } from '../../../models/repository'
+import { RepositoriesStore } from '../repositories-store'
+import { Branch, BranchType } from '../../../models/branch'
+import { GitStoreCache } from '../git-store-cache'
+import { getMergedBranches, deleteBranch } from '../../git'
+import { fatalError } from '../../fatal-error'
+import { RepositoryStateCache } from '../repository-state-cache'
+import * as moment from 'moment'
+
+/** Check if a repo needs to be pruned at least every 4 hours */
+const BackgroundPruneMinimumInterval = 1000 * 60 * 60 * 4
+const ReservedBranches = [
+  'master',
+  'gh-pages',
+  'HEAD',
+  'develop',
+  'dev',
+  'development',
+  'trunk',
+  'devel',
+  'release',
+]
+
+export class BranchPruner {
+  private timer: number | null = null
+
+  public constructor(
+    private readonly repository: Repository,
+    private readonly gitStoreCache: GitStoreCache,
+    private readonly repositoriesStore: RepositoriesStore,
+    private readonly repositoriesStateCache: RepositoryStateCache,
+    private readonly onPruneCompleted: (repository: Repository) => Promise<void>
+  ) {}
+
+  public async start() {
+    if (this.timer !== null) {
+      fatalError(
+        `A background prune task is already active and cannot begin pruning on ${
+          this.repository.name
+        }`
+      )
+    }
+
+    await this.pruneLocalBranches()
+    this.timer = window.setInterval(
+      () => this.pruneLocalBranches(),
+      BackgroundPruneMinimumInterval
+    )
+  }
+
+  public stop() {
+    if (this.timer === null) {
+      return
+    }
+
+    clearInterval(this.timer)
+    this.timer = null
+  }
+
+  private async findBranchesMergedIntoDefaultBranch(
+    repository: Repository,
+    defaultBranch: Branch
+  ): Promise<ReadonlyArray<string> | null> {
+    const gitStore = this.gitStoreCache.get(repository)
+    return (
+      (await gitStore.performFailableOperation(() =>
+        getMergedBranches(repository, defaultBranch.name)
+      )) || null
+    )
+  }
+
+  private async pruneLocalBranches(): Promise<void> {
+    if (this.repository.gitHubRepository === null) {
+      return
+    }
+
+    // Get the last time this repo was pruned
+    const lastPruneDate = await this.repositoriesStore.getLastPruneDate(
+      this.repository
+    )
+
+    // Only prune if it's been at least 24 hours since the last time
+    const dateNow = moment()
+    const threshold = dateNow.subtract(24, 'hours')
+
+    if (lastPruneDate !== null && threshold.isBefore(lastPruneDate)) {
```

Comment:
I was able to trip the wrong path because I got a `lastPruneDate` of `undefined` in here the first time I switched to a repository when running this up - could the new field have been missing from Dexie, despite `Date | null` being the expected type?

<img width="775" alt="screen shot 2019-02-06 at 10 43 43 am" src="https://user-images.githubusercontent.com/359239/52349173-54a35b00-29fc-11e9-96e0-b7483e101b2e.png">

Reply:
This eventually came good for me, which I think is down to the `undefined`/`null` distinction:

<img width="664" alt="screen shot 2019-02-06 at 10 49 19 am" src="https://user-images.githubusercontent.com/359239/52349488-0d699a00-29fd-11e9-9214-e82d78095e97.png">

The confusing part of this is the log messages:

```
info: [ui] Last prune took place in a day - skipping
```

Not sure whether you want to relax this check to `== null` here for those repositories which haven't been updated to see the new `null`.